sdb/commands/stacks.py

```diff
@@ -0,0 +1,255 @@
+            if func.type_.kind != drgn.TypeKind.FUNCTION:
+                raise sdb.CommandError(
+                    self.name, "'{:s}' is not a function".format(args.function))
+
+        task_states = Stacks.TASK_STATES.values()
+        if args.tstate and not args.tstate in task_states:
+            raise sdb.CommandError(
+                self.name,
+                "'{:s}' is not a valid task state (acceptable states: {:s})".
+                format(args.tstate, ", ".join(task_states)))
+
+        if args.module and Stacks.find_module_memory_segment(
+                self.prog, args.module)[0] == -1:
+            raise sdb.CommandError(
+                self.name,
+                "module '{:s}' doesn't exist or isn't currently loaded".format(
+                    args.module))
+
+    def call(self, objs: Iterable[drgn.Object]) -> Iterable[drgn.Object]:
+        # pylint: disable=too-many-locals
+        # pylint: disable=too-many-branches
+
+        #
+        # As the exception explains the code that follows this statement only
+        # works for linux kernel targets (crash dumps or live systems). When
+        # support for userland is added we can factor the kernel code that
+        # follows into its own function and switch to the correct codepath
+        # depending on the target.
+        #
+        if not self.prog.flags & drgn.ProgramFlags.IS_LINUX_KERNEL:
+            raise sdb.CommandError(self.name,
+                                   "userland targets are not supported yet")
+        self.validate_args(self.prog, self.args)
+
+        mod_start, mod_end = -1, -1
+        if self.args.module:
+            mod_start, mod_size = Stacks.find_module_memory_segment(
+                self.prog, self.args.module)
+            assert mod_start != -1
+            mod_end = mod_start + mod_size
+
+        print("{:<18s} {:<16s} {:6s}".format("TASK_STRUCT", "STATE", "COUNT"))
+        print("=" * 42)
+
+        stack_map = {}
+        stack_counter = Counter()
+        task_aggr = defaultdict(list)
+
+        #
+        # For now this iterates over the initial namespace only. The drgn
+        # API does provide support for iterating over a specific namespace
+        # so functiality can be added where the `stacks` command is passed
+        # namespace `drgn.Object`s through the pipeline or a namespace ID
+        # through `argparse` and iterates over the tasks of those. This
+        # functionality has been left as future work.
+        #
+        for task in for_each_task(self.prog):
+            stack_key, stack_frames = "", []
+            for frame in self.prog.stack_trace(task):
+                stack_key += str(frame.pc)
+                stack_frames.append(frame)
+            stack_counter[stack_key] += 1
+            stack_map[stack_key] = stack_frames
+            task_aggr[stack_key].append(
+                (task, Stacks.task_struct_get_state(task)))
```

Comment:
Do we really need `Stacks.task_struct_get_state(task)` in the data structure, versus calling it when necessary below?  I think you'd end up with the same number of calls either way.